File: 2439. Minimize Maximum of Array/Solution.java
class Solution {
    public int minimizeArrayValue(int[] nums) {
        long left = 0, right = 1000000000;
        int n = nums.length;
        while (left < right) {
            long mid = (left + right) / 2, carry = 0;
            for (int i = n - 1; i >= 0; i--) {
                carry = Math.max(nums[i] + carry - mid, 0);
            }
            if (carry == 0) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return (int) left;
    }
}

File: 2439. Minimize Maximum of Array/Solution.py
class Solution:
    def minimizeArrayValue(self, nums: List[int]) -> int:
        left, right, n = 0, 1e9, len(nums)
        while left < right:
            mid, carry = (left + right) // 2, 0
            for i in range(n - 1, -1, -1):
                carry = max(0, carry + nums[i] - mid)
            if carry == 0:
                right = mid
            else:
                left = mid + 1
        return int(left)